Pandas dependency deprecation future warning, add `pandera[pandas]` extra

cosmicBboy opened the pull request:
Fixes #1847

This PR is laying the groundwork for an eventual deprecation of the top-level `pandera` module to access the pandas validation objects, like schemas, models, types, etc:

- The `pandas` and `numpy` dependencies have been removed from `pandera`'s core dependencies.
- The top-level module imports for these symbols have been moved to `pandera.pandas`.
- Minimum pandas and numpy versions (2.1.1, 1.24.4 respectively) are enforced at import time

`pandera` will import `from pandera.pandas import *` so as to maintain backwards compatibility. The deprecation date for removing this import is TBD. I'll follow-up with additional PRs to update docs and tests to use `import pandera.pandas as pa` as the recommended way of defining pandas dataframe schemas.

This PR also adds a `pandas` extra so that users have to explicitly have `pandas` installed in their environment to use pandera's pandas data validation objects.

### ⚠️ Breaking Change

If you relied on `pandera` to install `pandas` in your environment (which I guess could happen, but would be strange), you will now have to install `pandas` separately:

```bash
pip install pandas pandera
```

Or

```bash
pip install "pandera[pandas]"
```

cosmicBboy commented:
hey @lundybernard, what do you about if we merge this PR and release 0.24.0 to rip off the bandage, then follow-up with the additional changes you proposed re: cleaning up tests, ci, etc.

lundybernard commented:
I would like to see confirmation that pandera works as expected without pandas installed.  
A successful run of the tests/polars tests without pandas installed would be sufficient, 
or manual testing on a project that requires polars but not pandas.

My concern is that projects which requires polars, dask, or pyspark, could be broken, if the Pandera code fails when pandas is not installed.
I would not recommend releasing 0.24.0 until we can confirm that these projects are safe.

I checked the virtual envs on my machine, and the github actions tests for polars and dask, which should no longer require pandas.
Those environments still include it.
see: https://github.com/unionai-oss/pandera/actions/runs/13689067152/job/38278821442?pr=1926#logs

cosmicBboy commented:
@lundybernard I've updated the noxfile and ci tests to make sure unit tests for each extra runs (esp. polars, pyspark, etc unit tests run without pandas installed)

I introduced a new `tests/base` directory as a home for base functionality. It's pretty sparse in terms of actual tests (there's only one test file right now) but will invest more in these base tests after this PR is merged.

lundybernard commented:
Checking the recent CI tests, 
Pandas is installed in the BaseTest environment: https://github.com/unionai-oss/pandera/actions/runs/14201631296/job/39789735672?pr=1926#step:6:137
And in the Dataframe Extras: Polars test environment: https://github.com/unionai-oss/pandera/actions/runs/14203942468/job/39797129700?pr=1926#step:7:143

It's being added to the Nox virtual environments, probably pulled in as a 2nd order dependency from one of the testing dependencies.

cosmicBboy commented:
good catch! more changes coming...

cosmicBboy commented:
hey @lundybernard, in the spirit of not letting perfect be the enemy of good (enough), I'll plan on merging this PR soon. Gonna pour over it one more time, but please feel free to point out any issues you might find with it after the fact, I can make tickets for it and address it in subsequent PRs.
